cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/AppFabricServer.java

```diff
@@ -110,36 +102,31 @@ public class AppFabricServer extends AbstractIdleService {
   private CommonNettyHttpServiceFactory commonNettyHttpServiceFactory;
 
   /**
-   * Construct the AppFabricServer with service factory and cConf coming from guice injection.
+   * Construct the AppFabricServer with service factory and cConf coming from guice
+   * injection.
    */
   @Inject
   public AppFabricServer(CConfiguration cConf, SConfiguration sConf,
-                         DiscoveryService discoveryService,
-                         @Named(Constants.Service.MASTER_SERVICES_BIND_ADDRESS) InetAddress hostname,
-                         @Named(Constants.AppFabric.HANDLERS_BINDING) Set<HttpHandler> handlers,
-                         @Nullable MetricsCollectionService metricsCollectionService,
-                         ProgramRuntimeService programRuntimeService,
-                         RunRecordCorrectorService runRecordCorrectorService,
-                         ProgramRunStatusMonitorService programRunStatusMonitorService,
-                         ApplicationLifecycleService applicationLifecycleService,
-                         ProgramNotificationSubscriberService programNotificationSubscriberService,
-                         ProgramStopSubscriberService programStopSubscriberService,
-                         @Named("appfabric.services.names") Set<String> servicesNames,
-                         @Named("appfabric.handler.hooks") Set<String> handlerHookNames,
-                         AuditLogSubscriberService auditLogSubscriberService,
-                         CoreSchedulerService coreSchedulerService,
-                         CredentialProviderService credentialProviderService,
-                         NamespaceCredentialProviderService namespaceCredentialProviderService,
-                         ProvisioningService provisioningService,
-                         BootstrapService bootstrapService,
-                         SystemAppManagementService systemAppManagementService,
-                         TransactionRunner transactionRunner,
-                         RunRecordMonitorService runRecordCounterService,
-                         CommonNettyHttpServiceFactory commonNettyHttpServiceFactory,
-                         RunDataTimeToLiveService runDataTimeToLiveService,
-                         SourceControlOperationRunner sourceControlOperationRunner,
-                         RepositoryCleanupService repositoryCleanupService,
-                         OperationNotificationSubscriberService operationNotificationSubscriberService) {
+      DiscoveryService discoveryService,
+      @Named(Constants.Service.MASTER_SERVICES_BIND_ADDRESS) InetAddress hostname,
+      @Named(Constants.AppFabric.SERVER_HANDLERS_BINDING) Set<HttpHandler> handlers,
+      @Nullable MetricsCollectionService metricsCollectionService,
+      ProgramRuntimeService programRuntimeService,
```

Comment:
I am pretty sure this one should exist in only 1 copy. Having multiple of twill watchers watching twill state with in-memory state would lead to problems

Reply:
As discussed, in this PR I have moved the HTTP handlers for program, app, workflow to processors.

They can be moved back to the Appfabric server after the issue due to in-memory cache is fixed. Created Jira for follow up PR: [CDAP-21112](https://cdap.atlassian.net/browse/CDAP-21112) `Move HTTP handler from Appfabric processor to server after fixing ProgramRuntimeService and RunRecordMonitorService`